kani-compiler/src/kani_middle/transform/contracts.rs

```diff
@@ -232,8 +228,297 @@ impl AnyModifiesPass {
+                        .unwrap();
+                    (true, run.body().unwrap())
+                } else {
+                    // Not a contract annotated function
+                    (false, body)
+                }
+            }
+            RigidTy::Closure(def, _args) => {
+                if self.unused_closures.contains(def) {
+                    // Delete body and mark it as unreachable.
+                    let mut new_body = MutableBody::from(body);
+                    new_body.clear_body(TerminatorKind::Unreachable);
+                    (true, new_body.into())
+                } else {
+                    // Not a contract annotated function
+                    (false, body)
+                }
+            }
+            _ => {
+                /* static variables case */
+                (false, body)
+            }
         }
     }
 }
+
+impl FunctionWithContractPass {
+    /// Build the pass by collecting which functions we are stubbing and which ones we are
+    /// verifying.
+    pub fn new(tcx: TyCtxt, unit: &CodegenUnit) -> FunctionWithContractPass {
+        if let Some(harness) = unit.harnesses.first() {
+            let attrs = KaniAttributes::for_instance(tcx, *harness);
+            let check_fn = attrs.interpret_for_contract_attribute().map(|(_, def_id, _)| def_id);
+            let replace_fns: HashSet<_> = attrs
+                .interpret_stub_verified_attribute()
+                .iter()
+                .map(|(_, def_id, _)| *def_id)
+                .collect();
+            FunctionWithContractPass { check_fn, replace_fns, unused_closures: Default::default() }
+        } else {
+            // Building the model for tests or public functions.
+            FunctionWithContractPass::default()
+        }
+    }
+
+    /// Functions with contract have the following structure:
+    /// ```ignore
+    /// fn original([self], args*) {
+    ///    let kani_contract_mode = kani::internal::mode(); // ** Replace this call
+    ///    match kani_contract_mode {
+    ///        kani::internal::RECURSION_CHECK => {
+    ///            let closure = |/*args*/|{ /*body*/};
+    ///            kani_register_contract(closure) // ** Replace this call
+    ///        }
+    ///        kani::internal::REPLACE => {
+    ///            // same as above
+    ///        }
+    ///        kani::internal::SIMPLE_CHECK => {
+    ///            // same as above
+    ///        }
+    ///        _ => { /* original code */}
+    ///    }
+    /// }
+    /// ```
+    /// See function `handle_untouched` inside `kani_macros`.
+    ///
+    /// Thus, we need to:
+    /// 1. Initialize `kani_contract_mode` variable to the value corresponding to the mode.
+    ///
+    /// Thus replace this call:
+    /// ```ignore
+    ///    let kani_contract_mode = kani::internal::mode(); // ** Replace this call
+    /// ```
+    /// by:
+    /// ```ignore
+    ///    let kani_contract_mode = mode_const;
+    ///    goto bbX;
```

Comment:
Why this `goto bbX`?

Reply:
Because we're replacing a terminator. Unless we want to add logic to merge the basic blocks, we need to insert a new terminator (e.g. a goto) that gets us to the next basic block that used to follow the function call.